src/torch/README.md

```diff
@@ -0,0 +1,131 @@
+
+These classes use custom `torch.autograd.Function` implementations (`_Operator`, `_ObjectiveFunction`, and `_ObjectiveFunctionGradient`) to define the forward and backward passes, handling the conversions between PyTorch tensors and SIRF objects.
+
+### `_Operator` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor to a SIRF object.
+    2.  Applies the SIRF `Operator.forward()` method.
+    3.  Converts the result back to a PyTorch tensor.
+    4.  If the input tensor requires gradients, it saves relevant information (the output SIRF object and the operator) in the context (`ctx`) for use in the backward pass.
+
+*   **Backward Pass (VJP):**
+    1.  Receives the "upstream gradient" (`grad_output`).
+    2.  Converts `grad_output` to a SIRF object.
+    3.  Applies the SIRF `Operator.backward()` method. This will apply the **Jacobian adjoint** of the operator to upstream gradient (the vector).
+    4.  Converts the resulting SIRF object back to a PyTorch tensor and returns it.
+
+### `_ObjectiveFunction` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor (representing an image) to a SIRF object.
+    2.  Calls the SIRF `ObjectiveFunction.get_value()` method.
+    3.  Returns the of the objective function value as a PyTorch tensor.
+    4. Saves relevant information to the `ctx` if gradients are needed.
+
+*   **Backward Pass (VJP):**
+    1.  Receives the upstream gradient (`grad_output`), in this case it is always a scalar.
+    2.  Gets the gradient of the objective function using `sirf_obj_func.get_gradient()`, which computed at the input and multiplied by the upstream gradient.
+    3.  Converts the SIRF gradient to a PyTorch tensor.
+    4.  Returns the gradient multiplied by `grad_output`.
+
+
+### `_ObjectiveFunctionGradient` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor to a SIRF object.
+    2.  Computes the *gradient* of the SIRF objective function using `sirf_obj_func.get_gradient()`, which is computed on the input.
+    3.  Returns the gradient as a PyTorch tensor.
+
+*   **Backward Pass (VJP):**
+    1.  Receives the upstream gradient (`grad_output`), which now represents a *vector* (not a scalar).
+    2.  Converts `grad_output` to a SIRF object.
+    3.  Multiples the Hessian evaluated as the input with the upstream gradient using `sirf_obj_func.multiply_with_Hessian()`.
+    4. Returns the Hessian multiplied with a vector.
+
+## Usage and Use Cases
+
+The `SIRFTorchOperator`, `SIRFTorchObjectiveFunction`, and `SIRFTorchObjectiveFunctionGradient` classes are designed to be used as standard PyTorch `nn.Module`s.  You would initialise them with the appropriate SIRF objects and then use them in your forward pass like any other PyTorch layer.
+
+The following use cases illustrate how to leverage this wrapper for different image reconstruction scenarios.
```

Comment:
as discussed, best to point to `use_cases.py`

Reply:
added a todo